compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeNarrower.java

```diff
@@ -221,20 +223,33 @@ public SymbolEnv evaluateTruth(BLangExpression expr, BLangNode targetNode, Symbo
      * @param expr Expression to evaluate
      * @param targetNode node to which the type narrowing applies
      * @param env Current environment
+     * @param isBinaryExpr Indicates whether the current context is a binary expression
      * @return target environment
      */
-    public SymbolEnv evaluateFalsity(BLangExpression expr, BLangNode targetNode, SymbolEnv env) {
+    public SymbolEnv evaluateFalsity(BLangExpression expr, BLangNode targetNode, SymbolEnv env, boolean isBinaryExpr) {
         Map<BVarSymbol, NarrowedTypes> narrowedTypes = getNarrowedTypes(expr, env);
         if (narrowedTypes.isEmpty()) {
             return env;
         }
 
         SymbolEnv targetEnv = getTargetEnv(targetNode, env);
-        narrowedTypes.forEach((symbol, typeInfo) -> {
-            BVarSymbol originalSym = getOriginalVarSymbol(symbol);
-            symbolEnter.defineTypeNarrowedSymbol(expr.pos, targetEnv, originalSym, typeInfo.falseType,
+        for (Map.Entry<BVarSymbol, NarrowedTypes> narrowedType : narrowedTypes.entrySet()) {
+            BType falseType = narrowedType.getValue().falseType;
+            BType trueType = narrowedType.getValue().trueType;
+
+            BVarSymbol originalSym = getOriginalVarSymbol(narrowedType.getKey());
+
+            if (isBinaryExpr) {
+                falseType = falseType == symTable.semanticError ?
+                        types.getRemainingType(originalSym.type, trueType) : falseType;
+            } else {
+                falseType = falseType == symTable.nullSet ? symTable.neverType : falseType;
+            }
+
+            symbolEnter.defineTypeNarrowedSymbol(expr.pos, targetEnv, originalSym,
+                    falseType == symTable.semanticError ? symTable.neverType : falseType,
```

Comment:
Can `falseType` still be `symTable.semanticError`?

Reply:
```ballerina
enum E { X, Y, Z }

function foo(E e) {
    if e is X {
    } else if e is Y {
    } else if e is Z { // when evaluating the falsity for the `else` node from this `if` node, the `falseType` is still `symTable.semanticError` since we handle the finite types separately in `getRemainingType()` method when narrowing. 
    } else if e is Y {
    }
}
```